src/services/htmlSelectionRange.ts

```diff
@@ -4,14 +4,21 @@
  *--------------------------------------------------------------------------------------------*/
 
 import { createScanner } from '../parser/htmlScanner';
-import { parse, Node } from '../parser/htmlParser';
+import { Node, HTMLParser, HTMLDocument } from '../parser/htmlParser';
 import { TokenType, Range, Position, SelectionRange, TextDocument } from '../htmlLanguageTypes';
 
-export function getSelectionRanges(document: TextDocument, positions: Position[]): SelectionRange[] {
+export class HTMLSelectionRange {
 
-	function getSelectionRange(position: Position): SelectionRange {
-		const applicableRanges = getApplicableRanges(document, position);
-		let prev : [number, number] | undefined = undefined;
+	constructor(private htmlParser: HTMLParser) {
+	}
+
+	public getSelectionRanges(document: TextDocument, positions: Position[]): SelectionRange[] {
+		const htmlDocument = this.htmlParser.parseDocument(document);
```

Comment:
this is a slight tweak to parse the document here, it used to be parsed inside `getApplicableRanges` which would happen for each position, now it only parses the document once.